subsys/nfc/ndef/le_oob_rec_parser.c

```diff
@@ -0,0 +1,368 @@
+					     u32_t payload_len,
+					     u8_t *result_buf,
+					     u32_t *result_buf_len)
+{
+	int err = 0;
+	const u32_t total_buf_size = *result_buf_len;
+	u32_t index = 0;
+	struct nfc_ndef_le_oob_rec_payload_desc *le_oob_rec_desc;
+	struct net_buf_simple mem_alloc;
+
+	if (!payload_buf) {
+		return -EINVAL;
+	}
+
+	/* Initialize memory allocator with buffer memory block and its size. */
+	net_buf_simple_init_with_data(&mem_alloc, result_buf, total_buf_size);
+	net_buf_simple_reset(&mem_alloc);
+
+	/* Allocate memory for LE OOB Record descriptor and initialize it. */
+	le_oob_rec_desc = (struct nfc_ndef_le_oob_rec_payload_desc *)
+		memory_allocate(&mem_alloc, sizeof(*le_oob_rec_desc));
+	if (!le_oob_rec_desc) {
+		return -ENOMEM;
+	}
+	memset(le_oob_rec_desc, 0, sizeof(*le_oob_rec_desc));
+
+	/* Parse record payload. */
+	while (index < payload_len) {
+		u8_t ad_len;
+
+		ad_len = payload_buf[index];
+		index += AD_LEN_FIELD_SIZE;
+
+		if ((ad_len == AD_LEN_INC_VAL) ||
+		    (index + ad_len > payload_len)) {
+			return -EINVAL;
+		}
+
+		u8_t ad_type = payload_buf[index];
+		const u8_t *ad_data =
+			&payload_buf[index + AD_TYPE_FIELD_SIZE];
+		u8_t ad_data_len = ad_len - AD_TYPE_FIELD_SIZE;
+
+		/* Decode AD payload. */
+		switch (ad_type) {
+		/* Mandatory fields. */
+		case BT_DATA_LE_BT_DEVICE_ADDRESS:
+			le_oob_rec_desc->addr = (bt_addr_le_t *)
+				memory_allocate(&mem_alloc,
+					sizeof(*le_oob_rec_desc->addr));
+			err = ble_device_address_decode(ad_data, ad_data_len,
+				le_oob_rec_desc->addr);
+			break;
+		case BT_DATA_LE_ROLE:
+			le_oob_rec_desc->le_role = (enum nfc_ndef_le_oob_rec_le_role *)
+				memory_allocate(&mem_alloc,
+					sizeof(*le_oob_rec_desc->le_role));
+			err = le_role_decode(ad_data, ad_data_len,
+				le_oob_rec_desc->le_role);
+			break;
+
+		/* Optional fields */
+		case BT_DATA_LE_SC_CONFIRM_VALUE:
+		case BT_DATA_LE_SC_RANDOM_VALUE:
+			if (!le_oob_rec_desc->le_sc_data) {
+				le_oob_rec_desc->le_sc_data = (struct bt_le_oob_sc_data *)
+					memory_allocate(&mem_alloc,
+						sizeof(*le_oob_rec_desc->le_sc_data));
+			}
+
+			if (ad_type == BT_DATA_LE_SC_CONFIRM_VALUE) {
+				err = le_sc_confirm_value_decode(ad_data,
+					ad_data_len,
+					le_oob_rec_desc->le_sc_data);
+			} else {
+				err = le_sc_random_value_decode(ad_data,
+					ad_data_len,
+					le_oob_rec_desc->le_sc_data);
+			}
+			break;
+		case BT_DATA_GAP_APPEARANCE:
+			le_oob_rec_desc->appearance = (u16_t *)
+				memory_allocate(&mem_alloc,
+					sizeof(*le_oob_rec_desc->appearance));
+			err = appearance_decode(ad_data, ad_data_len,
+				le_oob_rec_desc->appearance);
+			break;
+		case BT_DATA_FLAGS:
+			le_oob_rec_desc->flags = (u8_t *)
+				memory_allocate(&mem_alloc,
+					sizeof(*le_oob_rec_desc->flags));
+			err = flags_decode(ad_data, ad_data_len,
+				le_oob_rec_desc->flags);
+			break;
+		case BT_DATA_NAME_SHORTENED:
+		case BT_DATA_NAME_COMPLETE:
+		{
+			u8_t *name;
+
+			name = memory_allocate(&mem_alloc, ad_data_len + 1);
```

Comment:
Why is the `+1` needed? Can you use a define?

Reply:
BLE GAP Device name is stored in NDEF without NULL terminator:

```name[ad_data_len] = 0;```

I need to add it manually so that it's easy to print it out. Also, LE OOB Record encoder requires terminator.

Sure, I can add a definition for it.